dapps-certification-persistence/src/IOHK/Certification/Persistence/API.hs

```diff
@@ -1,16 +1,89 @@
+{-# LANGUAGE OverloadedLabels    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE OverloadedLabels #-}
-{-# LANGUAGE RecordWildCards #-}
-{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeOperators       #-}
 
 module IOHK.Certification.Persistence.API where
 
-import Database.Selda
-import Database.Selda.SQLite
-import IOHK.Certification.Persistence.Structure
-import Data.Maybe
-import Control.Monad
+import           Control.Monad
+import           Data.Maybe
+import           Database.Selda
+import           Database.Selda.SQLite
+import           IOHK.Certification.Persistence.Structure
+
+getTransactionIdQ:: Text -> Query t (Col t (ID Transaction))
+getTransactionIdQ  externalAddress = do
+  p <- select transactions
+  restrict (p ! #wtxExternalId .== text externalAddress)
+  pure (p ! #wtxId)
+
+getTransactionId :: MonadSelda m => Text -> m (Maybe (ID Transaction))
+getTransactionId = fmap listToMaybe . query . getTransactionIdQ
```

Comment:
I don't know `MonadSelda`, does it not have a way to express that a given `Query` has 1 or 0 results?

Reply:
From what I've seen, they are also using the `Maybe` type when a relation/computation could produce or not a value. For example when `leftJoin` is used the value representing the joined table in the `Query` is `Row t (Maybe a)` where `a` is the record representing the joined table. 
For example: 
in 
```haskell
getProfileQ :: ID Profile -> Query t (Row t Profile :*: Row t (Maybe DApp))
getProfileQ pid = do
  profile <- select profiles
  dapp <- leftJoin  (\dapp -> dapp ! #dappId .== literal pid) (select dapps)
  restrict (profile ! #profileId .== literal pid)
  pure (profile :*: dapp)
```

the `dapp` type is `Row t (Maybe Dapp)`